### ISSUE TITLE
Typescript definitions

### ISSUE THREAD

**Issue by Cyberuben** [enquirer/enquirer#135]
The Typescript definitions for this project seem to be lacking. I've seen a PR that has been closed because apparently it was "poorly written" but other than that I can't really find issues regarding types.

I'd love to use this library but right now I'm quite limited with the missing type definitions. I need the definitions for pre-defining questions (prompts?) before actually calling the methods.

**Comment by sw-yx** [comment]
i mean.. wanna PR it?

**Comment by jpike88** [comment]
![Screen Shot 2019-06-07 at 4 53 38 pm](https://user-images.githubusercontent.com/9585787/59096341-d07ac900-8944-11e9-89ec-da7446aa60e3.png)

**Comment by unional** [comment]
I'm taking a shot at this and publish a series of video/streaming along the way,
so that people can learn and understand how to add types to an existing library.

Also, it help explaining the decisions I make and the problems I encounter.

https://www.youtube.com/playlist?list=PLTLdF9P_MLntLaJt4JEDfT-6Fsr-j2O1Y

I noticed there are some inconsistency between the prompts and some seemingly bugs along the way.
I'll add them to this issue so that we can discuss what is the best approach to handle them.

I push my changes to the master branch of my fork: https://github.com/unional/enquirer

**Comment by unional** [comment]
It is still a work in progress.
I hope I can finish it in the next two weeks.

One question, are you open to convert the project to TypeScript? 🌷 
I'm not pushing for it or anything, but it will probably easier and cleaner if we do that.

**Comment by unional** [comment]
One of the inconsistency is that some of the common functions such as `render()` is returning `void` or `Promise<void>` depends on which prompt we can working with. My guess is that this is a bug and they all should be asynchronous.

`AutoComplete.space()` is taking a `ch character but the base class does not, so there could be a bug there:
```js
  space(ch) {
    // TODO: `super.space()` does not take param
    return this.options.multiple ? super.space(ch) : this.append(ch);
  }
```

Also in AutoComplete, the `pointer()` function has no param while be base class has, so it violates Liskov Substitution Principle.
```js
  // TODO: base type signature incompatible
  pointer() {
    return '';
  }
```

In `StringPrompt.keypress()`, it calls `append()` with two params while it only takes one:
```js
  async keypress(input, key = {}) {
    let prev = this.state.prevKeypress;
    this.state.prevKeypress = key;
    if (this.options.multiline === true && key.name === 'return') {
      if (!prev || prev.name !== 'return') {
        // TODO: `append()` takes 1 param
        return this.append('\n', key);
      }
    }
    return super.keypress(input, key);
  }
  append(ch) {
    let { cursor, input } = this.state;
    this.input = `${input}`.slice(0, cursor) + ch + `${input}`.slice(cursor);
    this.moveCursor(String(ch).length);
    this.render();
  }
```

In `Prompt.sections()`, it calls `validate()` with two params, but it only take one:
```js
    // TODO: `this.validate()` does not take two arguments
    let result = this.state.error || await this.validate(this.value, this.state);
...
  validate(value) {
    if (this.options.required === true) {
      return this.isValue(value);
    }
    return true;
  }
```

**Comment by unional** [comment]
As I'm doing this, I'm also adding more tests to cover all use cases. At the end of the exercise, the cover coverage should be very close to 100%.

**Comment by doowb** [comment]
Hi @unional thank you for taking this on! There was [another conversation](https://github.com/enquirer/enquirer/issues/212) about `types` here that you might want to take a look at.

> One question, are you open to convert the project to TypeScript? 🌷

For now, we'd prefer to keep this project JavaScript (as mentioned in the other issue, we have a refactor in the works and would like to complete that before making a change like converting to TypeScript).

If you're able to get the current typings up to date, and possibly add a way to lint/test the typings when changes are made to the code, that would be much appreciated :) If it would be easier to add the typings to `DefinitelyTyped` and keep them up-to-date through that project, we're open to that also since it decouples any typing languages from this project directly.

> As I'm doing this, I'm also adding more tests to cover all use cases. At the end of the exercise, the cover coverage should be very close to 100%.

😀 this popped up just as I was able to hit "Comment". Thank you, and I appreciate you doing this.

**Comment by unional** [comment]
Hi @doowb, no problem. :)
The DT approach wouldn't work because the library already expose its type.

If someone creates and uses `@types/enquirer` now, it will result in conflict.

So the solution is to either remove the current `index.d.ts` from distribution and user uses the typings form DT, or fix the `index.d.ts`.

> If you're able to get the current typings up to date, and possibly add a way to lint/test the typings when changes are made to the code, that would be much appreciated :)

What I have done is convert the existing tests from JavaScript to TypeScript.
All tests are use cases and they should work in TypeScript.

That's also why some of us who worked on `typings` in the past before TypeScript 2.0 likes to work on separate type repositories in the `@types` github organization because we can reuse the existing tests and also actually run and execute them.


I looked at the typings proposed here: https://github.com/enquirer/enquirer/pull/82 and it is not my cup of tea. It added a lot of extra stuff that "pollute" the export landscape (i.e. you get a lot of cruft when doing autocomplete on `import { ... } from 'enquirer'`).

What I have done in my own library and this typings is to maintain a clean interface. For example, currently it look like this:
![image](https://user-images.githubusercontent.com/3254987/72279784-b5afb780-35eb-11ea-925d-bd29ad4cf29c.png)

(there are still those `Choice` and `ChoiceOptions`, which I'll clean up along the way).

Another thing I would like to achieve is automatically detect the result type if possible.

Here is an example of what I mean:
https://github.com/unional/clibuilder/blob/master/src/cli/createCli.spec.ts#L208-L238

The `args` param is typed automatically based on the input in `arguments` and `options`.

Do you have a discussion thread for the refactoring?
I'll like to subscribe to it to keep up to date. 🌷

**Comment by doowb** [comment]
> If someone creates and uses `@types/enquirer` now, it will result in conflict.
> 
> So the solution is to either remove the current `index.d.ts` from distribution and user uses the typings form DT

I think when we do a major version bump, we'll want to do this (move the typings to `@types/enquirer`). After seeing the issue with `NodeJS.EventEmitter`, it would have been better to have the types external to the main code.

> What I have done is convert the existing tests from JavaScript to TypeScript.
> All tests are use cases and they should work in TypeScript.
> 
> That's also why some of us who worked on `typings` in the past before TypeScript 2.0 like to work on separate type repositories in the `@types` github organization because we can reuse the existing tests and also actually run and execute them.

I think this comment reinforces the last point of moving the types to the `DefinitelyTyped` org. I appreciate what you are doing, but I don't think the tests should be converted from JavaScript to TypeScript. This adds a barrier to maintainers that are not familiar with TypeScript and I believe will make it less likely that someone new will update tests if adding or fixing source code.

When I made the comment about linting/testing, I had forgotten that we already had a simple typescript test file that just ensured the definitions would work (e.g. with the correct syntax). I think this is all we need until we either migrate the entire project to TypeScript or move the definitions to DefinitelyTyped.

If converting the tests makes it easier on you while updating the type definitions, would you be able to make a PR for just the type definition changes?

Thanks again for working on these! I'm learning something about TypeScript 😉

**Comment by unional** [comment]
Not a problem~.
I convert the test from JavaScript to TypeScript because I can "eat my own dog food" as working on the type definition.
When everything is said and done, they can easily convert back to JavaScript. 🌷

**Comment by martinmcwhorter** [comment]
I have created a PR to start making the definitions more idiomatic with ESM imports https://github.com/enquirer/enquirer/pull/252

This is a breaking change, as `import * Enquirer from 'enquirer';` would become `import Enquirer from 'enquirer';`

This change would allow us to move to destructured ESM imports. If you are using, or would like to use, TypeScript with Enquirer -- please take a look at this PR.

**Comment by jpike88** [comment]
Throwing my five cents in: please don’t rely on @types for type definitions, it’s worth the minimal effort to maintain your own types:

- @types will likely be out of date at some point and some people will have a rude surprise
- type awareness helps boost refactoring efforts, types FTW

**Comment by unional** [comment]
@jpike88 100% agree to what you said about `@types`.
But in general it is the same when the typings are defined but not align to the code within the project.

@doowb, I've encountered many inconsistency between different ways to use `enquirer` so it makes adding the types to `enquirer` quite difficult.
And it can easily get out of sync when any of these behavior changes or fixed.

For example:
```ts
import Enquirer from 'enquirer'

Enquirer.prompt({
  type: 'input',
  name: 'xxx',
  header: 'header',
  message: 'message',
  footer: 'footer'
})

new Enquirer.Input({
  type: 'input',
  name: 'xxx',
  header: 'header',
  message: 'message',
  footer: 'footer'
})
```

The `header` and `footer` is not used in the first call but used in the second call.

How is your refactoring effort going?
I'm considering to get a smaller set of types ready for PR and improve upon it when the refactoring is done.

Or maybe you are open to try out TypeScript? :)
That would be my recommended approach.

**Comment by cpmech** [comment]
`Confirm` and `AutoComplete` are missing from index.d.ts. Temporary solution:

```
import Enquirer from 'enquirer';
  const response = await Enquirer.prompt(...)
  const confirm = new (Enquirer as any).Confirm(...)
  const complete = new (Enquirer as any).AutoComplete(...)
```

**Comment by rxliuli** [comment]
Is there any progress? The ts type is still problematic, especially when I am writing a custom prompt

**Comment by rxliuli** [comment]
> I could try fixing this sometime. What do you think?

我尝试了两种方式

1. Modify the enquirer to refactor part of it into ts, and use rollup to package - it has been abandoned. There are many codes that will cause ts errors. For the completed part, see: <https://github.com/rxliuli/enquirer/commit/115091ece950fa84be9b69653e93c2cdda642d2a>
  
![image](https://user-images.githubusercontent.com/24560368/117752152-e25bf980-b248-11eb-9b72-9dde69310faf.png)

2. Overwrite the index.d.ts of the enquirer locally, which is currently modified from the following version:<https://github.com/unional/enquirer/blob/typings/index.d.ts>，at least, you can now customize the plug-in correctly

![image](https://user-images.githubusercontent.com/24560368/117752654-bd1bbb00-b249-11eb-828e-296bd4a2eea3.png)

In any case, I can use [patch-package](https://www.npmjs.com/package/patch-package) to modify the enquirer in the project, if the enquirer does not or cannot be repaired

**Comment by rxliuli** [comment]
> Couldn't we just rewrite enquirer in typescript?

Yes, but it takes time to fix various problems, including but not limited to the following common codes in js

1. Rewrite the variable type, such as reassigning item in the for loop (not paying attention to the type change)
2. Use an imperative way to dynamically add fields or methods to the class
3. Overly flexible API, such as supporting various types of parameters, automatically binding this and the like
4. Use `{}` to assign a value to any object and then gradually assign each field below
5. More. . .

**Comment by rxliuli** [comment]
After taking a look, the project seems to be no longer actively maintained (no new version has been released for more than a year, and there are a lot of problems). Even if it seems that it is smaller than the inquirer and has a better object-oriented api (rather than a large and comprehensive super-complex function), it is still unacceptable to no longer maintain it. . .

**Comment by marcosdipaolo** [comment]
Any nice declaration file to share?

**Comment by vdegenne** [comment]
@unional How can I use your types? Do you provide them in one package?

**Comment by unional** [comment]
> @unional How can I use your types? Do you provide them in one package?

It's in the PR branch. But given it is few years ago, I'm not sure if it is still good or not.

For myself I have not been using `enquirer` since.

**Comment by vdegenne** [comment]
@unional Yes I see thanks for your work anyway, I moved on `@inquirer/prompts` and i'm satisfied with it for now
